### PR TITLE
test: `@cypherock/cysync-utils` package

### PR DESCRIPTION
**Pull request by jhasuraj01**
Ticket: [PRF-7100](https://app.clickup.com/t/9002019994/PRF-7100)

## Coverage Report (Istanbul)
- [x] 100% Statements 300/300
- [x] 100% Branches 141/141
- [x]  100% Functions 72/72
- [x] 100% Lines 271/271

File | Statements | - | Branches | - | Functions | - | Lines | - 
-- | -- | -- | -- | -- | -- | -- | -- | --
assert.ts | 100% | 5/5 | 100% | 6/6 | 100% | 1/1 | 100% | 5/5
bigNumber.ts | 100% | 74/74 | 100% | 6/6 | 100% | 28/28 | 100% | 69/69
config.ts | 100% | 2/2 | 100% | 0/0 | 100% | 0/0 | 100% | 2/2
getEnv.ts | 100% | 11/11 | 100% | 9/9 | 100% | 1/1 | 100% | 8/8
http.ts | 100% | 27/27 | 100% | 26/26 | 100% | 2/2 | 100% | 24/24
index.ts | 100% | 10/10 | 100% | 0/0 | 100% | 0/0 | 100% | 10/10
isSubset.ts | 100% | 5/5 | 100% | 0/0 | 100% | 2/2 | 100% | 3/3
logger.ts | 100% | 45/45 | 100% | 33/33 | 100% | 12/12 | 100% | 35/35
promiseQueue.ts | 100% | 37/37 | 100% | 11/11 | 100% | 8/8 | 100% | 34/34
resourceLock.ts | 100% | 70/70 | 100% | 48/48 | 100% | 14/14 | 100% | 68/68
sleep.ts | 100% | 3/3 | 100% | 0/0 | 100% | 2/2 | 100% | 3/3

### CODE REVIEW

**Comment by rohanagarwal94** [comment]
Task linked: [PRF-7100 Test: @cypherock/cysync-utils package](https://app.clickup.com/t/9002019994/PRF-7100)

File: packages/utils/src/promiseQueue.ts [hunk 1]
```diff
@@ -44,12 +48,15 @@ export class PromiseQueue<T> {
...
+      if (this.isAborted) return;
 
       if (this.runNext()) {
         const promiseFunc = this.todo.shift();
 
-        if (!promiseFunc) break;
+        if (!promiseFunc) {
+          await Promise.allSettled(this.running);
+          return;
+        }
```
Comment (irshadCypherock) [inline packages/utils/src/promiseQueue.ts]: Due to `return`,`onComplete` function call will not be triggred

File: packages/utils/src/promiseQueue.ts [hunk 2]
```diff
@@ -44,12 +48,15 @@ export class PromiseQueue<T> {
 
   async run() {
     while (this.total !== this.completed) {
-      if (this.isAborted) break;
+      if (this.isAborted) return;
```
Comment (irshadCypherock) [inline packages/utils/src/promiseQueue.ts:51]: - This will not trigger `onComplete`
- Check if `onComplete` is required if aborted

**Review by irshadCypherock — CHANGES_REQUESTED** [review]

File: packages/utils/src/promiseQueue.ts (hunk 2 above)
Comment (jhasuraj01) [reply packages/utils/src/promiseQueue.ts:51]: `onComplete` is not needed if `abort` is called

File: packages/utils/src/promiseQueue.ts (hunk 1 above)
Comment (jhasuraj01) [reply packages/utils/src/promiseQueue.ts]: `onComplete` is not needed if `abort` is called

File: packages/utils/src/comparator.ts [hunk 3]
```diff
@@ -2,6 +2,26 @@ import { BigNumber } from './bigNumber';
 
 export const createComparator =
```
Comment (irshadCypherock) [inline packages/utils/src/comparator.ts]: Looks like this function is not used anywhere, we can remove it.

File: packages/utils/tests/bigNumber/__fixtures__/valid.ts [hunk 4]
```diff
@@ -0,0 +1,640 @@
...
+      method: BigNumber.min,
+    },
+    {
+      name: 'maximum',
+      method: BigNumber.max,
+    },
+    {
+      name: 'minimum',
+      method: BigNumber.min,
+    },
```
Comment (irshadCypherock) [inline packages/utils/tests/bigNumber/__fixtures__/valid.ts]: Duplicate

File: packages/utils/tests/config.ts [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+describe('config', () => {
+  test('should export config object', async () => {
+    expect(config).toBeDefined();
+  });
+  test('should have LOG_LEVEL property', async () => {
+    expect(config.LOG_LEVEL).toBeDefined();
+  });
+  test('should have LOG_LEVEL property with default value', async () => {
+    expect(config.LOG_LEVEL).toEqual('debug');
+  });
```
Comment (irshadCypherock) [inline packages/utils/tests/config.ts]: Add test case: Config should change depending on the environment variable.

File: packages/utils/tests/logger.ts [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+  });
+
+  describe('createDefaultConsoleLogger', () => {
+    test('should log messages with meta based on log level priority', () => {
+      const actualConsole = global.console;
+
+      console.info = jest.fn();
+      console.debug = jest.fn();
+      console.warn = jest.fn();
+      console.error = jest.fn();
```
Comment (irshadCypherock) [inline packages/utils/tests/logger.ts:57]: This should be in `beforeEach` block

File: packages/utils/tests/logger.ts [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+      });
+      expect(console.error).toHaveBeenCalledWith(error.message, {
+        ...error.option,
+        ...logDefaultMeta,
+      });
+
+      console.info = actualConsole.info;
+      console.debug = actualConsole.debug;
+      console.warn = actualConsole.warn;
+      console.error = actualConsole.error;
```
Comment (irshadCypherock) [inline packages/utils/tests/logger.ts]: Should be in `afterEach` block

File: packages/utils/tests/logger.ts [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+    });
+
+    test('should log messages without meta based on log level priority', () => {
+      const infoLogger = jest.spyOn(console, 'info');
```
Comment (irshadCypherock) [inline packages/utils/tests/logger.ts]: Use `spyOn` or `jest.fn()` to monitor console function calls. Here we are using both methods.

**Review by irshadCypherock — CHANGES_REQUESTED** [review]

File: packages/utils/tests/bigNumber/__fixtures__/valid.ts (hunk 4 above)
Comment (jhasuraj01) [reply packages/utils/tests/bigNumber/__fixtures__/valid.ts]: removed duplication and added missing static methods.

File: packages/utils/tests/logger.ts (hunk 6 above)
Comment (jhasuraj01) [reply packages/utils/tests/logger.ts:57]: moved to beforeEach

File: packages/utils/tests/logger.ts (hunk 7 above)
Comment (jhasuraj01) [reply packages/utils/tests/logger.ts]: moved to after each

File: packages/utils/tests/logger.ts (hunk 8 above)
Comment (jhasuraj01) [reply packages/utils/tests/logger.ts]: using `jest.fn()`

File: packages/utils/tests/config.ts (hunk 5 above)
Comment (jhasuraj01) [reply packages/utils/tests/config.ts]: added new test

File: packages/utils/src/comparator.ts (hunk 3 above)
Comment (jhasuraj01) [reply packages/utils/src/comparator.ts]: removed `createComparator`

**Review by irshadCypherock — APPROVED** [review]

**Pull request merged**